File: js/combination_sum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function (candidates, target) {
  function backtrack(start, candidates, target, curr, res, currSum) {
    if (start >= candidates.length || currSum === target) {
      res.push([...curr]);
      return;
    }
    for (let i = start; i < candidates.length; ++i) {
      currSum += candidates[i];
      if (currSum <= target) {
        curr.push(candidates[i]);
        backtrack(start, candidates.slice(i), target, curr, res, currSum);
        curr.pop();
      }
      currSum -= candidates[i];
    }
  }
  const res = [];
  backtrack(0, candidates, target, [], res, 0);
  return res;
};

// algomonster solution
function combinationSum(candidates, target) {
  function dfs(nums, start, remaining, path) {
    if (remaining == 0) {
      res.push([...path]);
      return;
    }
    for (let i = start; i < nums.length; i++) {
      const num = nums[i];
      if (remaining - num < 0) continue;
      path.push(num);
      dfs(nums, i, remaining - num, path);
      path.pop();
    }
  }
  res = [];
  dfs(candidates, 0, target, []);
  return res;
}

File: python/combination_sum.py
from typing import List

def combination_sum(candidates: List[int], target: int) -> List[List[int]]:
    # WRITE YOUR BRILLIANT CODE HERE
    result = []
    def backtrack(curr, currSum):
        if currSum == target:
            sortedList = list(sorted(curr[:]))
            if sortedList not in result:
                result.append(sortedList)
            return
        
        for candidate in candidates:
            if currSum + candidate <= target:
                curr.append(candidate)
                backtrack(curr, currSum + candidate)
                curr.pop()
    
    backtrack([], 0)
    return result


    # solution

    # The way we dedup is to only use candidate numbers whose index in the array is >= last used number's index.
    # we can do this due to the DFS order
    # We use an additional state start_index to keep track of the position of the last used number 
from typing import List

def combination_sum(candidates: List[int], target: int) -> List[List[int]]:
    def dfs(nums, start_index, remaining, path):
        if remaining == 0:
            res.append(path[:])
            return
        for i in range(start_index, len(nums)):
            num = nums[i]
            if remaining - num < 0:
                continue
            dfs(nums, i, remaining - num, path + [num])
    res = []
    dfs(candidates, 0, target, [])
    return res